[CALCITE-5948] Use explicit casting if element type in ARRAY/MAP does not equal derived component type

chucheng92 opened the pull request:
https://issues.apache.org/jira/browse/CALCITE-5948

tanclary commented on core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java:
```diff
@@ -1309,6 +1311,64 @@ public static boolean isMeasure(SqlNode selectItem) {
+      } else {
+        elementType = oddType;
+      }
+      if (operandTypes.get(i).equalsSansFieldNames(elementType)) {
```
can this be rewritten as `if(!operandTypes.get(i)....) { call.setOperand() }`

chucheng92 replied:
yes, we can. fixed. thanks tanner for reviewing

tanclary commented on core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java:
```diff
@@ -1309,6 +1311,64 @@ public static boolean isMeasure(SqlNode selectItem) {
   }
 
+  /**
+   * When the array element does not equal with the array component type, making explicit casting.
```
i think you can remove `with` and change making->make

chucheng92 replied:
thanks. fixed.

chucheng92 commented:
@tanclary hi, tanner. I have solved some comments. If you have time, could you help to review it again? thanks!

chucheng92 commented:
@tanclary hi, Tanner, If you have time, would you please review it again? 

This issue provides the public method of element explicit cast of array/map, if it can be solved in time, it will also be of great benefit to subsequent related issues.

Looking forward your comments/reviews.

chucheng92 commented on core/src/test/resources/sql/misc.iq:
```diff
@@ -2167,12 +2167,12 @@ select array[1,null,2] as a from (values (1));
++------------------+
+| EXPR$0           |
++------------------+
+| [a  , null, bcd] |
```
@JiajunBernoulli thanks for reviewing. this is a import change in this PR.

if we have `array('A', 'AB')`, the component type is `char(2)`, because the element explicit casting bug we not cast 'A' to `char(2)` to match derived component type. So if we fix this bug to match the `char(2)`, it means 'A' need cast to `char(2)`, it will fill will extra spaces by calcite default.

This is calcite and sql standard limitation. Of course, this does look weird, actually calcite has other ways to remove spaces. We can activate `SqlConformanceEnum.PRAGMATIC_2003` to let `shouldConvertRaggedUnionTypesToVarying` be true, it will change array/map char type to varchar type to skip these spaces. And I also noticed that some engines such as apache flink set `shouldConvertRaggedUnionTypesToVarying` to true by default. (calcite is false by default)

More details can be found in javadoc https://github.com/apache/calcite/blob/d9dd3ac8a9f695e111a0a5e77f45b61b90f4b5b6/core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java#L465C7-L465C7

I also have added some cases in the PR to show this difference.

chucheng92 replied:
@JiajunBernoulli hi, Jiajun. If you have time, could you help to review it again?

JiajunBernoulli replied:
Ok, Thank you for your clear explanation.

tanclary reviewed (approved):
Hi @chucheng92 Sorry for delay, thanks for addressing the changes. I'm okay to merge this as soon as @JiajunBernoulli is too.

chucheng92 commented:
Thanks Tanner and Jiajun for reviewing. All comments are resolved/confirmed. 
I have rebased main and squashed the commits and made commit name to match Jira name. 
hi @tanclary If you have time, could you help to merge this?

chucheng92 commented:
@julianhyde hi, julian. I have noticed you give a comment in the ticket to suggest give a test case for calcite-5960. I have added it, if you have time, could you help to review it? If you approve that, then i will squash the commits.

vlsi commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java:
```diff
@@ -973,6 +973,9 @@ private static Expression scaleIntervalToNumber(
       RelDataType sourceType,
       RelDataType targetType,
       Expression operand) {
+    if (targetType.getSqlTypeName().getFamily() == null) {
+      return operand;
+    }
     switch (requireNonNull(targetType.getSqlTypeName().getFamily(),
         () -> "SqlTypeFamily for " + targetType)) {
```
Please rework, and remove code duplication.
Extract `family` into a regular variable, and perform a regular null check. Then `requireNonNull` will not be needed.

Thank you so much for doing so.

chucheng92 replied:
@vlsi thanks for reviewing. agree with you. I have fixed it. If you have time, please take a look. thanks

chucheng92 replied:
@vlsi hi, considering [calcite-5960](https://issues.apache.org/jira/browse/CALCITE-5960) has been resolved. Now this PR has nothing to do with calcite-5960. thanks again.

chucheng92 commented:
> @julianhyde hi, julian. I noticed that you gave a comment in the ticket to suggest giving a test case for [calcite-5960](https://issues.apache.org/jira/browse/CALCITE-5960). I have added it, if you have time, could you help to review it? If you approve that, then i will squash the commits. thanks!

calcite-5960 has been resolved.

chucheng92 commented:
thanks Julian and Vladimir for reviewing the related commit calcite-5960. Considering that commit has been resolved, so I have rebased this PR. 

Currently all comments are resolved. 

Hi, @tanclary sorry to ping you, if you have time, could you help to recheck/merge this? 
thank you all again for patient reviewing.

tanclary commented on babel/src/test/resources/sql/big-query.iq:
```diff
@@ -1012,9 +1012,9 @@ FROM
 
 SELECT
```
Nit: could we change the commit message to "Use explicit casting if element type in ARRAY/MAP does not equal derived component type"

chucheng92 replied:
thanks, both updated the Jira and commit name to this.

tanclary commented on core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java:
```diff
@@ -1309,6 +1311,63 @@ public static boolean isMeasure(SqlNode selectItem) {
+    RelDataType elementType;
+    for (int i = 0; i < operands.size(); i++) {
+      if (i % 2 == 0) {
+        elementType = evenType;
```
could you explain `even` and `odd` types? I am not sure what that means

chucheng92 replied:
Yes, updated the PR to explain these 2 parameters and gave a example about the usage of this method.

chucheng92 commented on core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java:
```diff
@@ -1309,6 +1311,63 @@ public static boolean isMeasure(SqlNode selectItem) {
+    }
+  }
+
+  private static SqlNode castTo(SqlNode node, RelDataType type) {
```
updated.

chucheng92 commented:
hi, @tanclary I have added javadoc in these private methods with a single commit, If you have time, PTAL again. 

thanks and sorry to bother you.

chucheng92 commented:
hi @tanclary Tanner, sorry to ping you, I have added javadoc in these 2 private methods by a single commit, If you have time, could you help to review it again?  thanks!

chucheng92 commented:
hi Tanner @tanclary , if you are free, could you help to take a look again? I have some follow-up commit/bugfix may depend on this. Sorry to bother you.

chucheng92 commented:
@tanclary Hi, Tanner, I have updated the commit/jira name and add javadocs. Since it's been a long time, If you have time, can you help to review it again? thank you.

tanclary reviewed (approved):
Hi @chucheng92 sorry for the delay, I was out unexpectedly for a couple of weeks. Thank you for resolving all of the comments, I think this looks great.

chucheng92 commented on core/src/test/java/org/apache/calcite/test/JdbcTest.java:
```diff
@@ -8144,10 +8144,9 @@ private void checkGetTimestamp(Connection con) throws SQLException {
-    // Result should be "EXPR$0=[1, 1.1]\n"; [CALCITE-4850] logged.
     CalciteAssert.that()
         .query("select array[1, 1.1]")
-        .returns("EXPR$0=[0E+1, 1.1]\n");
```
@tanclary thanks for patient reviewing! btw, I think we can safely close CALCITE-4850. The result has been corrected.
